--- 
title: "Fish eDNA"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib]
biblio-style: apalike
link-citations: yes
github-repo: rstudio/bookdown-demo
---

# Introduction {-}



## Objectives {-}



## Dataset {-}

The real data set consists of 18 water samples obtained from rivers and lakes near oil-sands extraction sites in the province of Alberta, Canada. Water samples were filtered and DNA collected on the filters was extracted and sequenced. For each sample, two different marker genes were used to amplify and sequence distinct DNA regions : the 12S ribosomal RNA gene and the mitochondrial cytochrome oxidase subunit I (COI). 

```{r, echo = FALSE, warning = FALSE}
library(DT)
primer_info_table = read.csv(file = "data/primer_table.csv", check.names = FALSE)
datatable(primer_info_table, rownames = FALSE, caption="Amplified regions details", options=list(scrollX=T, dom = 't', paging=FALSE, ordering=FALSE,
                                                            initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '10px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '10px')
```

## Alberta fishes {-}

[Fishbase](https://www.fishbase.se/search.php) was used to generate the following list of Alberta freshwater fishes. 

```{r, echo = FALSE, warning = FALSE}
library(DT)
fishbase_alberta = read.csv(file = "data/fishbase_alberta_freswater.csv", check.names = FALSE)
datatable(fishbase_alberta, rownames = FALSE, options=list(initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '11px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '11px')
```

## Methods and their associated available databases {-}

Within the framework of this project, the following methods and their associated databases were evaluated in regards to their capacities of correctly identifying sequences belonging to freshwater fishes from Alberta.    

```{r, echo = FALSE, warning = FALSE}
library(DT)
workflow_database = read.csv(file = "data/workflow_databases.csv", check.names = FALSE)
datatable(workflow_database, rownames = FALSE, options=list(scrollX=T, dom = 't', paging=FALSE, ordering=FALSE,
                                                            initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '11px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '11px')
```

**Links** 

- [1] [Barcode of life (BOLD)](https://www.boldsystems.org/)
- [2] [barque workflow](https://github.com/enormandeau/barque) 
- [3] [COI formatted for Barque](https://www.ibis.ulaval.ca/services/bioinformatique/barque_databases/)
- [4] [DADA2 workflow](https://benjjneb.github.io/dada2/ITS_workflow.html)
- [5] [Eukaryote CO1 Classifier](https://github.com/terrimporter/CO1Classifier/releases/tag/RDP-COI-v5.1.0) 
- [6] [Mitochondrial Genome Database of Fish (MitoFish)](https://mitofish.aori.u-tokyo.ac.jp/) 
- [7] [National Center for Biotechnology Information (NCBI)](https://www.ncbi.nlm.nih.gov/)
- [8] [Ribosomal Database Project (RDP) classifier](https://github.com/rdpstaff/classifier)
- [9] [VSEARCH](https://github.com/torognes/vsearch)
- [10] [12S fish Classifier v1.0.1](https://github.com/terrimporter/12SfishClassifier/releases/tag/v1.0.1)
- [11] [12S formatted for barque](https://github.com/enormandeau/barque/blob/master/03_databases/12S.fasta.gz)

## Representation of Alberta freshwater fishes in databases {-}

Each database was queried to evaluate the presence of reference sequences for Alberta freshwater fishes. Code used to queried the databases can be found [here](https://github.com/karinevilleneuve/fish_eDNA/blob/master/scripts/01_representation_freshwater_fishes_db.R). 

```{r, echo = FALSE, message = FALSE, warning = FALSE}
#library(shiny)
library(DT)
library(dplyr)

aff_databases = read.csv(file = "data/represensation_albertafish_databases.csv", check.names = FALSE, row.names = 1)
datatable(aff_databases, rownames = FALSE, options = list(scrollX = T, dom = 't', paging=FALSE, ordering=FALSE,
                                                        initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '11px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '11px') %>% 
  formatStyle(names(aff_databases), 
              backgroundColor = styleEqual(c("Yes","No"), c("#DeF7E9", "#ff7f7f")), 
              backgroundSize = "50%", 
              backgroundPosition = 'center')
```

The absence of the following fishes from any databases can be attributed to different names being used. The presence of these other names in the databases was therefore also evaluated. 

```{r, echo = FALSE, message = FALSE, warning = FALSE}
#library(shiny)
library(DT)
library(dplyr)

aff_databases = read.csv(file = "data/missing_fishes.csv", check.names = FALSE, row.names = 1)
datatable(aff_databases, rownames = FALSE, options = list(scrollX = T, dom = 't', paging=FALSE, ordering=FALSE,
                                                        initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '11px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '11px') %>% 
  formatStyle(names(aff_databases), 
              backgroundColor = styleEqual(c("Yes","No"), c("#DeF7E9", "#ff7f7f")), 
              backgroundSize = "50%", 
              backgroundPosition = 'center')
```
