<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Evaluate performance of databases | Fish eDNA</title>
  <meta name="description" content="Evaluate performance of databases | Fish eDNA" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Evaluate performance of databases | Fish eDNA" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Evaluate performance of databases | Fish eDNA" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="taxonomic-classification.html"/>
<link rel="next" href="generating-custom-database-of-the-12s-marker-gene.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#objectives"><i class="fa fa-check"></i>Objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#dataset"><i class="fa fa-check"></i>Dataset</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#alberta-fishes"><i class="fa fa-check"></i>Alberta fishes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#methods-and-their-associated-available-databases"><i class="fa fa-check"></i>Methods and their associated available databases</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#representation-of-alberta-freshwater-fishes-in-databases"><i class="fa fa-check"></i>Representation of Alberta freshwater fishes in databases</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html"><i class="fa fa-check"></i>Required files, scripts and programs</a>
<ul>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#programs"><i class="fa fa-check"></i>Programs</a></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#databases"><i class="fa fa-check"></i>Databases</a></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#scripts"><i class="fa fa-check"></i>Scripts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html"><i class="fa fa-check"></i>Processing raw fastq</a>
<ul>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html#dada2"><i class="fa fa-check"></i>DADA2</a></li>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html#barque"><i class="fa fa-check"></i>barque</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-classification.html"><a href="taxonomic-classification.html"><i class="fa fa-check"></i>Taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html"><i class="fa fa-check"></i>Evaluate performance of databases</a>
<ul>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#description"><i class="fa fa-check"></i>Description</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#methods"><i class="fa fa-check"></i>Methods</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#implementing-cvi-with-our-data"><i class="fa fa-check"></i>Implementing CVI with our data</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#preformatting-the-databases"><i class="fa fa-check"></i>Preformatting the databases</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#generate-benchmark-datasets"><i class="fa fa-check"></i>Generate benchmark datasets</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#generate-predictions"><i class="fa fa-check"></i>Generate predictions</a>
<ul>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#vsearch"><i class="fa fa-check"></i>VSEARCH</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#rdp"><i class="fa fa-check"></i>RDP</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html"><i class="fa fa-check"></i>Generating custom database of the 12S marker gene</a>
<ul>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#evaluate-resolution-blindspots"><i class="fa fa-check"></i>Evaluate resolution blindspots</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#prepare-files-for-classifiers"><i class="fa fa-check"></i>Prepare files for classifiers</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#notes-on-resolution-blindspot"><i class="fa fa-check"></i>Notes on resolution blindspot</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#assessing-hybrids"><i class="fa fa-check"></i>Assessing hybrids</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html"><i class="fa fa-check"></i>Generating custom database</a>
<ul>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html#notes-on-resolution-blindspot-1"><i class="fa fa-check"></i>Notes on resolution blindspot</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html#assessing-hybrids-1"><i class="fa fa-check"></i>Assessing hybrids</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-mock-communities.html"><a href="generating-mock-communities.html"><i class="fa fa-check"></i>Generating mock communities</a></li>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html"><i class="fa fa-check"></i>Supplementary workflow</a>
<ul>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#curating-databases-for-distance-matrix"><i class="fa fa-check"></i>Curating databases for distance matrix</a>
<ul>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#rdp-1"><i class="fa fa-check"></i>RDP</a></li>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#barque-1"><i class="fa fa-check"></i>Barque</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Fish eDNA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="evaluate-performance-of-databases" class="section level1 unnumbered hasAnchor">
<h1>Evaluate performance of databases<a href="evaluate-performance-of-databases.html#evaluate-performance-of-databases" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Link to <a href="https://peerj.com/articles/4652/">article</a> and <a href="https://drive5.com/taxxi/doc/index.html">TAXXI website</a>.</p>
<div id="description" class="section level2 unnumbered hasAnchor">
<h2>Description<a href="evaluate-performance-of-databases.html#description" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Edgar (2018) developed a Cross-Validation by Identity (CVI) framework that tests 16 unique classifiers and 3 parameter settings of the SINTAX, RDP, and Non-Bayensien-Classifier classifiers to assign taxonomy to specially modified test and training data sets (Warcup ITS, 16S full length, 16S V4 and 16S V3-5). These data sets were designed by Edgar (2018) to 1) have even representation across genera, and 2) test classifier effectiveness across different loci of the same gene (16S). Each classifier is then assessed for the true positive rate (TPR), the over-classification rate (OCR), the under classification rate (UCR), the misclassification rate (MCR), and the accuracy (ACC).</p>
<ul>
<li>The true positive rate (TPR) indicates how frequently the correct taxonomy was assigned out of the total number of opportunities for correct classification.</li>
<li>The over-classification rate (OCR) indicates how frequently too many ranks are predicted for query sequences out of the total opportunities to make an over classification error.</li>
<li>The under-classification rate (UCR) indicates how frequently too few ranks are predicted for query sequences out of the total number of opportunities to make this error.</li>
<li>The misclassification rate (MCR) indicates how frequently a sequence matching a query is available in the database but is not predicted for that query out to the number of opportunities to make this error.</li>
<li>The accuracy (ACC) indicates the number of correct taxonomic calls out of the number of opportunities to determine correct taxonomy.</li>
</ul>
</div>
<div id="methods" class="section level2 unnumbered hasAnchor">
<h2>Methods<a href="evaluate-performance-of-databases.html#methods" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>A reference with known taxonomies is split into test and training sets such that for all test sequences, the most similar training sequence has a given identity (<em>d</em>). This is repeated for different identities, enabling assessment of prediction accuracy at varying distances from the reference.</p>
<p><img style="padding:30px" src="data/taxxi_figures.jpg" align="left" width="20%"/></p>
<p><br />
- <em>R</em> is the reference dataset divided into four disjoint subsets <em>S</em>, <em>T</em>, <em>W</em> and <em>Z</em>.<br />
- <em>S</em> is the <strong>test set</strong>.<br />
- <em>A</em> is the <strong>training set</strong> formed by the union of <em>T</em> and <em>W</em>.<br />
- <em>T</em> is the set of top hits for sequences in <em>S</em>, which are constrained to have identities in the range <em>d</em> ± σ<br />
(where σ specifies the maximum allowed deviation from the desired identity (<em>d</em>)).<br />
- <em>W</em> contains reference sequences with identity &lt; <em>d</em>; these are retained to create the largest possible training set.<br />
- <em>Z</em> contains sequences which cannot be assigned to <em>S</em>, <em>T</em> or <em>W</em> without violating the identity constraint.</p>
<p><br clear="left"/></p>
</div>
<div id="implementing-cvi-with-our-data" class="section level2 unnumbered hasAnchor">
<h2>Implementing CVI with our data<a href="evaluate-performance-of-databases.html#implementing-cvi-with-our-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Usearch was downloaded here : <a href="https://github.com/rcedgar/usearch_old_binaries/blob/main/bin/usearch11.0.667_i86linux64" class="uri">https://github.com/rcedgar/usearch_old_binaries/blob/main/bin/usearch11.0.667_i86linux64</a></p>
<p>TAXXI framework was used to compare the performance of the RDP and vsearch classifiers with the following databases :</p>
<p>12S.fasta =&gt; 12s_vsearch.fasta —— Evaluate with vsearch
12S_rdp.fasta =&gt; 12s_rdp.fasta —— Evaluate with RDP
12s_alberta_curated.fasta =&gt; 12s_cur.fasta — Evaluate with RDP and vsearch</p>
<p>bold.fasta =&gt; coi_vsearch.fasta —————————– Evaluate with vsearch
mytrainseq.fasta =&gt; coi_rdp.fasta —————————- Evaluate with RDP
coi_alberta_curated_derep_NCBI.fasta =&gt; coi_cur.fasta ——– Evaluate with RDP and vsearch</p>
<ul>
<li>COI BOLD <a href="https://www.ibis.ulaval.ca/services/bioinformatique/barque_databases/">3</a></li>
<li>COI Eukaryote V5.1.0 <a href="https://github.com/terrimporter/CO1Classifier/releases/tag/RDP-COI-v5.1.0">5</a></li>
<li>12S MitoFISH V1.0.1<br />
</li>
<li>Custom 12S from barque</li>
</ul>
</div>
<div id="preformatting-the-databases" class="section level2 unnumbered hasAnchor">
<h2>Preformatting the databases<a href="evaluate-performance-of-databases.html#preformatting-the-databases" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ol style="list-style-type: decimal">
<li>Make all sequence uppercase (RDP COI only) and if necessary replace tab for semi-colon in sequence header</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="evaluate-performance-of-databases.html#cb8-1" tabindex="-1"></a><span class="ex">/home/karine/seqkit</span> seq coi_rdp.fasta <span class="at">--upper-case</span> <span class="at">-o</span> coi_rdp_upp.fasta</span>
<span id="cb8-2"><a href="evaluate-performance-of-databases.html#cb8-2" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&#39;s/\t/;/g&#39;</span> database.fasta </span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Randomly select 10 sequences per species (COI databases only)</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="evaluate-performance-of-databases.html#cb9-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">&quot;&gt;&quot;</span> coi_rdp_upp.fasta <span class="op">&gt;</span> coi_rdp_header.txt</span>
<span id="cb9-2"><a href="evaluate-performance-of-databases.html#cb9-2" tabindex="-1"></a><span class="co"># Add unique sequence ID to BOLD database </span></span>
<span id="cb9-3"><a href="evaluate-performance-of-databases.html#cb9-3" tabindex="-1"></a><span class="fu">awk</span> <span class="st">&#39;/^&gt;/{sub(&quot;&gt;&quot;, &quot;&gt;&quot;++i&quot;_&quot;)}1&#39;</span> coi_vsearch.fasta <span class="op">&gt;</span> coi_vsearchID.fasta</span>
<span id="cb9-4"><a href="evaluate-performance-of-databases.html#cb9-4" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">&quot;&gt;&quot;</span> coi_vsearchID.fasta <span class="op">&gt;</span> coi_vsearch_header.txt</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="evaluate-performance-of-databases.html#cb10-1" tabindex="-1"></a><span class="do">#####################################################</span></span>
<span id="cb10-2"><a href="evaluate-performance-of-databases.html#cb10-2" tabindex="-1"></a><span class="do">######################## RDP ########################</span></span>
<span id="cb10-3"><a href="evaluate-performance-of-databases.html#cb10-3" tabindex="-1"></a><span class="do">#####################################################</span></span>
<span id="cb10-4"><a href="evaluate-performance-of-databases.html#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="evaluate-performance-of-databases.html#cb10-5" tabindex="-1"></a>raw_rdp <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;/home/kvilleneuve/fish_edna/code/evalute_db/coi_rdp_header.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;;&quot;</span>)</span>
<span id="cb10-6"><a href="evaluate-performance-of-databases.html#cb10-6" tabindex="-1"></a>rdp_10sp_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb10-7"><a href="evaluate-performance-of-databases.html#cb10-7" tabindex="-1"></a>i <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb10-8"><a href="evaluate-performance-of-databases.html#cb10-8" tabindex="-1"></a></span>
<span id="cb10-9"><a href="evaluate-performance-of-databases.html#cb10-9" tabindex="-1"></a><span class="cf">for</span> (species <span class="cf">in</span> <span class="fu">unique</span>(raw_rdp<span class="sc">$</span>V10)){</span>
<span id="cb10-10"><a href="evaluate-performance-of-databases.html#cb10-10" tabindex="-1"></a>  i <span class="ot">=</span> i <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb10-11"><a href="evaluate-performance-of-databases.html#cb10-11" tabindex="-1"></a>  sub_df <span class="ot">=</span> <span class="fu">subset</span>(raw_rdp, V10 <span class="sc">==</span> species)</span>
<span id="cb10-12"><a href="evaluate-performance-of-databases.html#cb10-12" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(sub_df) <span class="sc">&gt;</span> <span class="dv">10</span>) {</span>
<span id="cb10-13"><a href="evaluate-performance-of-databases.html#cb10-13" tabindex="-1"></a>    rand_sub_df <span class="ot">=</span> sub_df[<span class="fu">sample</span>(<span class="fu">nrow</span>(sub_df), <span class="at">size=</span><span class="dv">10</span>), ] </span>
<span id="cb10-14"><a href="evaluate-performance-of-databases.html#cb10-14" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb10-15"><a href="evaluate-performance-of-databases.html#cb10-15" tabindex="-1"></a>    rand_sub_df <span class="ot">=</span> sub_df</span>
<span id="cb10-16"><a href="evaluate-performance-of-databases.html#cb10-16" tabindex="-1"></a>  }</span>
<span id="cb10-17"><a href="evaluate-performance-of-databases.html#cb10-17" tabindex="-1"></a>  rdp_10sp_list[[i]] <span class="ot">=</span> rand_sub_df</span>
<span id="cb10-18"><a href="evaluate-performance-of-databases.html#cb10-18" tabindex="-1"></a>}</span>
<span id="cb10-19"><a href="evaluate-performance-of-databases.html#cb10-19" tabindex="-1"></a>rdP10sp <span class="ot">=</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, rdp_10sp_list)</span>
<span id="cb10-20"><a href="evaluate-performance-of-databases.html#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="evaluate-performance-of-databases.html#cb10-21" tabindex="-1"></a><span class="fu">write.table</span>(rdP10sp, </span>
<span id="cb10-22"><a href="evaluate-performance-of-databases.html#cb10-22" tabindex="-1"></a>            <span class="st">&quot;/home/kvilleneuve/fish_edna/code/evalute_db/coi_rdp_header2keep.txt&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;;&quot;</span>)</span>
<span id="cb10-23"><a href="evaluate-performance-of-databases.html#cb10-23" tabindex="-1"></a></span>
<span id="cb10-24"><a href="evaluate-performance-of-databases.html#cb10-24" tabindex="-1"></a><span class="do">#####################################################</span></span>
<span id="cb10-25"><a href="evaluate-performance-of-databases.html#cb10-25" tabindex="-1"></a><span class="do">######################## BOLD ########################</span></span>
<span id="cb10-26"><a href="evaluate-performance-of-databases.html#cb10-26" tabindex="-1"></a><span class="do">#####################################################</span></span>
<span id="cb10-27"><a href="evaluate-performance-of-databases.html#cb10-27" tabindex="-1"></a></span>
<span id="cb10-28"><a href="evaluate-performance-of-databases.html#cb10-28" tabindex="-1"></a>raw <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;/home/karine/coi_vsearch_header.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb10-29"><a href="evaluate-performance-of-databases.html#cb10-29" tabindex="-1"></a>raw<span class="sc">$</span>Species <span class="ot">=</span> <span class="fu">paste</span>(raw<span class="sc">$</span>V2, raw<span class="sc">$</span>V3, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb10-30"><a href="evaluate-performance-of-databases.html#cb10-30" tabindex="-1"></a>list_10sp <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb10-31"><a href="evaluate-performance-of-databases.html#cb10-31" tabindex="-1"></a>i <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb10-32"><a href="evaluate-performance-of-databases.html#cb10-32" tabindex="-1"></a></span>
<span id="cb10-33"><a href="evaluate-performance-of-databases.html#cb10-33" tabindex="-1"></a><span class="cf">for</span> (specie <span class="cf">in</span> <span class="fu">unique</span>(raw<span class="sc">$</span>Species)){</span>
<span id="cb10-34"><a href="evaluate-performance-of-databases.html#cb10-34" tabindex="-1"></a>  i <span class="ot">=</span> i <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb10-35"><a href="evaluate-performance-of-databases.html#cb10-35" tabindex="-1"></a>  sub_df <span class="ot">=</span> <span class="fu">subset</span>(raw, Species <span class="sc">==</span> specie)</span>
<span id="cb10-36"><a href="evaluate-performance-of-databases.html#cb10-36" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nrow</span>(sub_df) <span class="sc">&gt;</span> <span class="dv">10</span>) {</span>
<span id="cb10-37"><a href="evaluate-performance-of-databases.html#cb10-37" tabindex="-1"></a>    rand_sub_df <span class="ot">=</span> sub_df[<span class="fu">sample</span>(<span class="fu">nrow</span>(sub_df), <span class="at">size=</span><span class="dv">10</span>), ] </span>
<span id="cb10-38"><a href="evaluate-performance-of-databases.html#cb10-38" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb10-39"><a href="evaluate-performance-of-databases.html#cb10-39" tabindex="-1"></a>    rand_sub_df <span class="ot">=</span> sub_df</span>
<span id="cb10-40"><a href="evaluate-performance-of-databases.html#cb10-40" tabindex="-1"></a>  }</span>
<span id="cb10-41"><a href="evaluate-performance-of-databases.html#cb10-41" tabindex="-1"></a>  list_10sp[[i]] <span class="ot">=</span> rand_sub_df</span>
<span id="cb10-42"><a href="evaluate-performance-of-databases.html#cb10-42" tabindex="-1"></a>}</span>
<span id="cb10-43"><a href="evaluate-performance-of-databases.html#cb10-43" tabindex="-1"></a>final_10sp <span class="ot">=</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, list_10sp)</span>
<span id="cb10-44"><a href="evaluate-performance-of-databases.html#cb10-44" tabindex="-1"></a></span>
<span id="cb10-45"><a href="evaluate-performance-of-databases.html#cb10-45" tabindex="-1"></a>final_10sp<span class="sc">$</span>good_name <span class="ot">=</span> <span class="fu">paste</span>(final_10sp<span class="sc">$</span>V1, final_10sp<span class="sc">$</span>V4, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)</span>
<span id="cb10-46"><a href="evaluate-performance-of-databases.html#cb10-46" tabindex="-1"></a>final_10sp<span class="sc">$</span>good_name <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">&quot;&gt;&quot;</span>, <span class="st">&quot;&quot;</span>, final_10sp<span class="sc">$</span>good_name)</span>
<span id="cb10-47"><a href="evaluate-performance-of-databases.html#cb10-47" tabindex="-1"></a></span>
<span id="cb10-48"><a href="evaluate-performance-of-databases.html#cb10-48" tabindex="-1"></a>final <span class="ot">=</span> <span class="fu">as.data.frame</span>(final_10sp<span class="sc">$</span>good_name)</span>
<span id="cb10-49"><a href="evaluate-performance-of-databases.html#cb10-49" tabindex="-1"></a></span>
<span id="cb10-50"><a href="evaluate-performance-of-databases.html#cb10-50" tabindex="-1"></a><span class="fu">write.table</span>(final, </span>
<span id="cb10-51"><a href="evaluate-performance-of-databases.html#cb10-51" tabindex="-1"></a>            <span class="st">&quot;/home/karine/coi_vsearch_header2keep.txt&quot;</span>, </span>
<span id="cb10-52"><a href="evaluate-performance-of-databases.html#cb10-52" tabindex="-1"></a>            <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;;&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="evaluate-performance-of-databases.html#cb11-1" tabindex="-1"></a><span class="ex">./seqkit</span> grep <span class="at">-f</span> coi_vsearch_header2keep.txt coi_vsearch.fasta <span class="at">-o</span> coi_vsearch_10sp.fasta <span class="at">--threads</span> 40</span>
<span id="cb11-2"><a href="evaluate-performance-of-databases.html#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="evaluate-performance-of-databases.html#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="evaluate-performance-of-databases.html#cb11-4" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&#39;s/&gt;//g&#39;</span> coi_rdp_header2keep.txt</span>
<span id="cb11-5"><a href="evaluate-performance-of-databases.html#cb11-5" tabindex="-1"></a><span class="ex">/home/karine/seqkit</span> grep <span class="at">-f</span> coi_rdp_header2keep.txt coi_rdp_upp.fasta <span class="at">-o</span> coi_rdp_10sp.fasta <span class="at">--threads</span> 20</span></code></pre></div>
</div>
<div id="generate-benchmark-datasets" class="section level2 unnumbered hasAnchor">
<h2>Generate benchmark datasets<a href="evaluate-performance-of-databases.html#generate-benchmark-datasets" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>1. Generate distance matrix </strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="evaluate-performance-of-databases.html#cb12-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.fasta <span class="kw">;</span> <span class="cf">do</span> <span class="ex">/home/karine/usearch11</span> <span class="at">-calc_distmx</span> <span class="va">$i</span> <span class="at">-maxdist</span> 0.2 <span class="at">-termdist</span> 0.3 <span class="at">-tabbedout</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>_dist.txt<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p><strong>2. Generate training and test sets </strong></p>
<p>The <code>distmx_split_identity</code> command from <a href="https://www.drive5.com/usearch/manual/cvi.html">USEARCH</a> divides sequences into subsets such that the top-hit identity is a given value. This is used to create test-training pairs for cross-validation by identity. Input is a distance matrix created by the <code>calc_distmx command</code>. As per the methods specified in Edgar (2018) maximum allowed deviation (σ) from <em>d</em> used : σ = 1% for <em>d</em> = 90% and σ = 0.5% for <em>d</em> = 99, 97 and 95%. The following code was adapted from <a href="https://doi.org/10.1016/j.ecoinf.2024.102817">Donhauser <em>et al.</em> (2024)</a> and validated with available documentation from <a href="https://www.drive5.com/usearch/manual/cvi.html">USEARCH</a>.</p>
<p>To create training and test sets at different identity thresholds</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="evaluate-performance-of-databases.html#cb13-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.fasta <span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb13-2"><a href="evaluate-performance-of-databases.html#cb13-2" tabindex="-1"></a>        <span class="ex">~/usearch11</span> <span class="at">-distmx_split_identity</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>_dist.txt <span class="at">-mindist</span> 0.025 <span class="at">-maxdist</span> 0.035 <span class="at">-tabbedout</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.97.subsets.txt</span>
<span id="cb13-3"><a href="evaluate-performance-of-databases.html#cb13-3" tabindex="-1"></a>        <span class="ex">~/usearch11</span> <span class="at">-distmx_split_identity</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>_dist.txt <span class="at">-mindist</span> 0.045 <span class="at">-maxdist</span> 0.055 <span class="at">-tabbedout</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.95.subsets.txt</span>
<span id="cb13-4"><a href="evaluate-performance-of-databases.html#cb13-4" tabindex="-1"></a>        <span class="ex">~/usearch11</span> <span class="at">-distmx_split_identity</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>_dist.txt <span class="at">-mindist</span> 0.095 <span class="at">-maxdist</span> 0.105 <span class="at">-tabbedout</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.90.subsets.txt</span>
<span id="cb13-5"><a href="evaluate-performance-of-databases.html#cb13-5" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>Output is a tabbed text given by the -tabbedout option. Fields are:</p>
<ul>
<li>Col1 - <code>Subset name</code> : there are four subsets with names <code>1</code>, <code>2</code>, <code>1x</code> and <code>2x</code>.</li>
<li>Col2 - <code>Label1</code> : label1 is the label of a sequence in the subset given by Col1.</li>
<li>Col3 - <code>Label2</code> : label2 is the top hit in the other subset (1 or 2)</li>
<li>Col4 - <code>Dist</code> : distance between Label1 and Label2.</li>
</ul>
<p>Subsets <code>1</code> and <code>2</code> have top hits to each other in the specified range. Subset <code>1x</code> has lower identities with subset <code>2</code>, and can therefore be added to the training set if subset <code>2</code> is the query set. Similarly, subset <code>2x</code> has lower identities with subset <code>1</code> and can be added to the training set if subset <code>1</code> is the query</p>
<p>Get sequence ID for training and test set</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="evaluate-performance-of-databases.html#cb14-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.subsets.txt <span class="kw">;</span> <span class="cf">do</span> <span class="fu">awk</span> <span class="st">&#39;$1==1 || $1==&quot;1x&quot; {print $2}&#39;</span> <span class="va">$i</span> <span class="op">&gt;</span> <span class="va">$i</span>.trainingIDs.txt<span class="kw">;</span> <span class="fu">awk</span> <span class="st">&#39;$1==2 {print $2}&#39;</span> <span class="va">$i</span> <span class="op">&gt;</span> <span class="va">$i</span>.testIDs.txt<span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>Create FASTA file with test and training set at each identity, use subset 1 as training and subset 2 as test.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="evaluate-performance-of-databases.html#cb15-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.trainingIDs.txt <span class="kw">;</span> <span class="cf">do</span> <span class="ex">~/seqkit</span> grep <span class="at">-n</span> <span class="at">-f</span> <span class="va">$i</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.fasta <span class="op">&gt;</span> <span class="va">$i</span>.trainning.fasta <span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb15-2"><a href="evaluate-performance-of-databases.html#cb15-2" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.testIDs.txt <span class="kw">;</span> <span class="cf">do</span> <span class="ex">~/seqkit</span> grep <span class="at">-n</span> <span class="at">-f</span> <span class="va">$i</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.fasta <span class="op">&gt;</span> <span class="va">$i</span>.test.fasta <span class="kw">;</span> <span class="cf">done</span> </span></code></pre></div>
<p><strong>3. Rename files</strong></p>
<p>By renaming the files this way we can use a for loop to generate the predictions.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="evaluate-performance-of-databases.html#cb16-1" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> <span class="pp">*</span>subsets<span class="pp">*</span> <span class="kw">;</span> <span class="cf">do</span> <span class="fu">mv</span> <span class="va">$file</span> <span class="va">${file</span><span class="op">//</span><span class="ss">.9</span><span class="op">/</span>_9<span class="va">}</span> <span class="kw">;</span> <span class="cf">done</span></span></code></pre></div>
<p>We now have, for each identity level, a set for training and a set for testing. We can now generate the prediction with either VSEARCH or RDP</p>
</div>
<div id="generate-predictions" class="section level2 unnumbered hasAnchor">
<h2>Generate predictions<a href="evaluate-performance-of-databases.html#generate-predictions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="vsearch" class="section level3 unnumbered hasAnchor">
<h3>VSEARCH<a href="evaluate-performance-of-databases.html#vsearch" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Using option <code>usearch_global</code> (option used by barque) :</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="evaluate-performance-of-databases.html#cb17-1" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="pp">*</span>.subsets.txt.testIDs.txt.test.fasta <span class="kw">;</span> <span class="cf">do</span> <span class="ex">vsearch</span> <span class="at">--usearch_global</span> <span class="va">$i</span> <span class="at">-db</span> <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.subsets.txt.trainingIDs.txt.trainning.fasta <span class="at">--blast6out</span>  <span class="va">${i</span><span class="op">%%</span>.<span class="pp">*</span><span class="va">}</span>.vsearch_predictions.tsv <span class="at">--top_hits_only</span> <span class="at">--notrunclabels</span> <span class="at">--id</span> 0.70<span class="kw">;</span> <span class="cf">done</span> </span></code></pre></div>
<p>Output is a tabbed text given by the -blast6out option. Fields are:</p>
<ul>
<li>Col1 : <code>query_name</code></li>
<li>Col2 : <code>target</code> - database sequence label</li>
<li>Col3 : <code>id</code> - percentage of identity (real value ranging from 0.0 to 100.0). The percentage identity is defined as 100 * (matching columns) / (alignment length - terminal gaps). See fields id0 to id4 for other definitions.</li>
<li>Col4 : <code>alnlen</code></li>
<li>Col5 : <code>mism</code></li>
<li>Col6 : <code>opens</code></li>
<li>Col7 : <code>qlo</code></li>
<li>Col8 : <code>qhi</code></li>
<li>Col9 : <code>tlo</code></li>
<li>Col10 : <code>thi</code></li>
<li>Col11 : <code>evalue</code></li>
<li>Col12 : <code>bits</code></li>
</ul>
</div>
<div id="rdp" class="section level3 unnumbered hasAnchor">
<h3>RDP<a href="evaluate-performance-of-databases.html#rdp" class="anchor-section" aria-label="Anchor link to header"></a></h3>

</div>
</div>
</div>
<hr>
<center>
  <div class="footer">
      Karine Villeneuve
      <p : id="date"></p>
<script>
  // Get current date and time
//  var now = new Date();
//  var date = now.toLocaleString();
//  n =  new Date();
//  y = n.getFullYear();
//  m = n.getMonth();
//  d = n.getDate();
//  document.getElementById("date").innerHTML = d + "/" + m + "/" + y;
  n = new Date().toDateString()
  document.getElementById("date").innerHTML = n
  // Insert date and time into HTML
//  document.getElementById("date").innerHTML = date;
</script>
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="taxonomic-classification.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="generating-custom-database-of-the-12s-marker-gene.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/karinevilleneuve/fish_eDNA/master/04_evaluatedDB.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
