<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Generating custom database of the 12S marker gene | Fish eDNA</title>
  <meta name="description" content="Generating custom database of the 12S marker gene | Fish eDNA" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Generating custom database of the 12S marker gene | Fish eDNA" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Generating custom database of the 12S marker gene | Fish eDNA" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="evaluate-performance-of-databases.html"/>
<link rel="next" href="generating-custom-database.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link href="libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.33/datatables.js"></script>
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#objectives"><i class="fa fa-check"></i>Objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#dataset"><i class="fa fa-check"></i>Dataset</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#alberta-fishes"><i class="fa fa-check"></i>Alberta fishes</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#methods-and-their-associated-available-databases"><i class="fa fa-check"></i>Methods and their associated available databases</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#representation-of-alberta-freshwater-fishes-in-databases"><i class="fa fa-check"></i>Representation of Alberta freshwater fishes in databases</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html"><i class="fa fa-check"></i>Required files, scripts and programs</a>
<ul>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#programs"><i class="fa fa-check"></i>Programs</a></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#databases"><i class="fa fa-check"></i>Databases</a></li>
<li class="chapter" data-level="" data-path="required-files-scripts-and-programs.html"><a href="required-files-scripts-and-programs.html#scripts"><i class="fa fa-check"></i>Scripts</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html"><i class="fa fa-check"></i>Processing raw fastq</a>
<ul>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html#dada2"><i class="fa fa-check"></i>DADA2</a></li>
<li class="chapter" data-level="" data-path="processing-raw-fastq.html"><a href="processing-raw-fastq.html#barque"><i class="fa fa-check"></i>barque</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="taxonomic-classification.html"><a href="taxonomic-classification.html"><i class="fa fa-check"></i>Taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html"><i class="fa fa-check"></i>Evaluate performance of databases</a>
<ul>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#description"><i class="fa fa-check"></i>Description</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#methods"><i class="fa fa-check"></i>Methods</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#implementing-cvi-with-our-data"><i class="fa fa-check"></i>Implementing CVI with our data</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#preformatting-the-databases"><i class="fa fa-check"></i>Preformatting the databases</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#generate-benchmark-datasets"><i class="fa fa-check"></i>Generate benchmark datasets</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#generate-predictions"><i class="fa fa-check"></i>Generate predictions</a>
<ul>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#vsearch"><i class="fa fa-check"></i>VSEARCH</a></li>
<li class="chapter" data-level="" data-path="evaluate-performance-of-databases.html"><a href="evaluate-performance-of-databases.html#rdp"><i class="fa fa-check"></i>RDP</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html"><i class="fa fa-check"></i>Generating custom database of the 12S marker gene</a>
<ul>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#evaluate-resolution-blindspots"><i class="fa fa-check"></i>Evaluate resolution blindspots</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#prepare-files-for-classifiers"><i class="fa fa-check"></i>Prepare files for classifiers</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#notes-on-resolution-blindspot"><i class="fa fa-check"></i>Notes on resolution blindspot</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database-of-the-12s-marker-gene.html"><a href="generating-custom-database-of-the-12s-marker-gene.html#assessing-hybrids"><i class="fa fa-check"></i>Assessing hybrids</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html"><i class="fa fa-check"></i>Generating custom database</a>
<ul>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html#notes-on-resolution-blindspot-1"><i class="fa fa-check"></i>Notes on resolution blindspot</a></li>
<li class="chapter" data-level="" data-path="generating-custom-database.html"><a href="generating-custom-database.html#assessing-hybrids-1"><i class="fa fa-check"></i>Assessing hybrids</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="generating-mock-communities.html"><a href="generating-mock-communities.html"><i class="fa fa-check"></i>Generating mock communities</a></li>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html"><i class="fa fa-check"></i>Supplementary workflow</a>
<ul>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#curating-databases-for-distance-matrix"><i class="fa fa-check"></i>Curating databases for distance matrix</a>
<ul>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#rdp-1"><i class="fa fa-check"></i>RDP</a></li>
<li class="chapter" data-level="" data-path="supplementary-workflow.html"><a href="supplementary-workflow.html#barque-1"><i class="fa fa-check"></i>Barque</a></li>
</ul></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Fish eDNA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="generating-custom-database-of-the-12s-marker-gene" class="section level1 unnumbered hasAnchor">
<h1>Generating custom database of the 12S marker gene<a href="generating-custom-database-of-the-12s-marker-gene.html#generating-custom-database-of-the-12s-marker-gene" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Sequences from the RDP databases consist of the complete 12S region while the sequences from the barque databases are trimmed to the amplified region of interest. Therefore, I am keeping the sequences from the RDP databases and adding to it some missing sequences.</p>
<p>Steps overview</p>
<ol style="list-style-type: decimal">
<li>Filter RDP database</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li>Extract sequence header</li>
<li>Keep only sequence header of fish found in Alberta</li>
<li>Curate database using the curated sequence names</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>Add missing sequences</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li>Download sequences from Genbank</li>
<li>Modify sequence header</li>
<li>Combine with curated RDP database</li>
</ol>
<ol style="list-style-type: decimal">
<li><strong>Filter RDP database</strong></li>
</ol>
<ol style="list-style-type: lower-alpha">
<li><strong>Extract sequence header</strong></li>
</ol>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb18-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">&quot;&gt;&quot;</span> 12S_rdp.fasta <span class="op">&gt;</span> header_12s_rdp.txt </span></code></pre></div>
<ol start="2" style="list-style-type: lower-alpha">
<li><strong>Keep only sequence header of fish found in Alberta</strong></li>
</ol>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-1" tabindex="-1"></a><span class="do">### ------------------------- Load the libraries -------------------------------------------------- #### </span></span>
<span id="cb19-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb19-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-3" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb19-4"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-4" tabindex="-1"></a><span class="do">### ------------------------- Save path for output -------------------------------------------------- #### </span></span>
<span id="cb19-5"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-5" tabindex="-1"></a>path <span class="ot">=</span> <span class="st">&quot;/home/kvilleneuve/fish_edna/database&quot;</span></span>
<span id="cb19-6"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-6" tabindex="-1"></a><span class="do">### ------------------------- Load list of Alberta freshwater fish -------------------------------------------------- #### </span></span>
<span id="cb19-7"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-7" tabindex="-1"></a>alberta_fish <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">&quot;/home/kvilleneuve/fish_edna/database/fishbase_alberta_with_missing.csv&quot;</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-8"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-8" tabindex="-1"></a>species <span class="ot">=</span> <span class="fu">gsub</span>(<span class="st">&quot; &quot;</span>, <span class="st">&quot;_&quot;</span>, alberta_fish<span class="sc">$</span>Species) <span class="co"># Replace space with underscore </span></span>
<span id="cb19-9"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-9" tabindex="-1"></a><span class="do">### ------------------------- Load database sequence header -------------------------------------------------- #### </span></span>
<span id="cb19-10"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-10" tabindex="-1"></a><span class="co"># I am intentionally specifying the wrong separator because I don&#39;t want to separate the columns yet</span></span>
<span id="cb19-11"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-11" tabindex="-1"></a>rdp12s_raw <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;/home/kvilleneuve/fish_edna/database/header_rdp_12S.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>)</span>
<span id="cb19-12"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-12" tabindex="-1"></a>filtered <span class="ot">=</span> rdp12s_raw <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="dv">1</span>, str_detect, <span class="fu">paste0</span>(species, <span class="at">collapse =</span> <span class="st">&#39;|&#39;</span>)))</span>
<span id="cb19-13"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb19-13" tabindex="-1"></a><span class="fu">write.table</span>(filtered, <span class="at">file =</span> <span class="fu">paste</span>(path, <span class="st">&quot;/alberta_ID_&quot;</span>, <span class="fu">names</span>(list_db)[i], <span class="st">&quot;.txt&quot;</span>, <span class="at">sep =</span><span class="st">&quot;&quot;</span>), <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<ol start="3" style="list-style-type: lower-alpha">
<li><strong>Curate database using the curated sequence names</strong></li>
</ol>
<p>Use <code>seqkit</code> tool to filter databases in order to keep only sequences of freshwater fishes found in Alberta using list of ID generated for each database.</p>
<p>For RDP</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb20-1" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&#39;s/\t/;/g&#39;</span> 12S_rdp.fasta</span>
<span id="cb20-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb20-2" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&#39;s/\t/;/g&#39;</span> alberta_ID_RDP_12S.txt</span>
<span id="cb20-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb20-3" tabindex="-1"></a><span class="ex">seqkit</span> grep <span class="at">-nrf</span> alberta_ID_RDP_12S.txt 12S_rdp.fasta <span class="at">-o</span> 12s_01_alberta_curated_RDP_raw.fasta</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li><strong>Add missing sequences</strong></li>
</ol>
<ol style="list-style-type: lower-alpha">
<li><strong>Download sequences from Genbank</strong></li>
</ol>
<p>Missing sequences were downloaded from NCBI</p>
<ol start="2" style="list-style-type: lower-alpha">
<li><strong>Modify sequence header</strong></li>
</ol>
<p>Sequence header of downloaded sequences were manually modified to include all ranks.</p>
<ol start="3" style="list-style-type: lower-alpha">
<li><strong>Combine with curated RDP database</strong></li>
</ol>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb21-1" tabindex="-1"></a><span class="co"># 12S </span></span>
<span id="cb21-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb21-2" tabindex="-1"></a><span class="fu">cat</span> 12s_01_alberta_curated_RDP_raw.fasta 12s_02_missing_sequences.fasta <span class="op">&gt;</span>  12s_03_alberta_curated_withmissing.fasta</span>
<span id="cb21-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb21-3" tabindex="-1"></a></span>
<span id="cb21-4"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb21-4" tabindex="-1"></a><span class="co"># Normalize length </span></span>
<span id="cb21-5"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb21-5" tabindex="-1"></a><span class="ex">seqkit</span> seq <span class="at">-w</span> 60 12s_03_alberta_curated_withmissing.fasta <span class="op">&gt;</span> 12s_04_alberta_curated_clean.fasta</span></code></pre></div>
<div id="evaluate-resolution-blindspots" class="section level2 unnumbered hasAnchor">
<h2>Evaluate resolution blindspots<a href="generating-custom-database-of-the-12s-marker-gene.html#evaluate-resolution-blindspots" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To evaluate resolution blindspot between two or more taxa the the sequence from the databases were trimmed using the Forward and Reverse primer (rev-comp) with Cutadapt. The sequences were then dereplicated using vsearch <code>derep_fulllength</code> and every sequences removed were evaluated by aligning them with MEGA11.</p>
<p>Trim</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb22-1" tabindex="-1"></a><span class="co"># Forward primer </span></span>
<span id="cb22-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb22-2" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">-g</span> <span class="at">--action</span><span class="op">=</span>retain CCGGTAAAACTCGTGCCAGC <span class="at">-o</span> 12s_alberta_curated_combined_trimF.fasta 12s_03_alberta_curated_withmissing.fasta <span class="at">-e</span> 0.2 <span class="at">--untrimmed-output</span> readswithnoforward.txt</span>
<span id="cb22-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb22-3" tabindex="-1"></a><span class="co"># Reverse primer</span></span>
<span id="cb22-4"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb22-4" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">-a</span> CAAACTGGGATTAGATACCCCACTATG <span class="at">-o</span> 12s_alberta_curated_combined_trimFR.fasta  12s_alberta_curated_combined_trimF.fasta</span></code></pre></div>
<p>Dereplicate</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb23-1" tabindex="-1"></a><span class="ex">vsearch</span> <span class="at">--derep_fulllength</span> 12s_alberta_curated_combined_trimFR.fasta <span class="at">--output</span> 12s_alberta_curated_combined_trim_derep.fasta</span></code></pre></div>
<p>See section <a href="Notes%20on%20resolution%20blindspot">Notes on resolution blindspot</a> for more details about which sequences were removed.</p>
</div>
<div id="prepare-files-for-classifiers" class="section level2 unnumbered hasAnchor">
<h2>Prepare files for classifiers<a href="generating-custom-database-of-the-12s-marker-gene.html#prepare-files-for-classifiers" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>::::::::::::::: vsearch Classifier for barque :::::::::::::::::</p>
<ol style="list-style-type: decimal">
<li>Copy the curate database to folder <code>03_database</code> in the barque directory</li>
<li>Rename file with shorter name</li>
<li>gzip file</li>
</ol>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb24-1" tabindex="-1"></a><span class="fu">cp</span> /home/kvilleneuve/fish_edna/database/curated12s_db/12s_04_alberta_curated_clean.fasta /home/kvilleneuve/fish_edna/code/barque_12s/03_databases/12Scurated.fasta</span>
<span id="cb24-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb24-2" tabindex="-1"></a><span class="fu">gzip</span> 12Scurated.fasta</span></code></pre></div>
<p>::::::::::::::: RDP Classifier :::::::::::::::::</p>
<p><strong>Prepare taxonomy file</strong></p>
<ol style="list-style-type: decimal">
<li>Extract sequence header using grep (2) Replace semi-colon with tab (3) remove &gt;</li>
</ol>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb25-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb25-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-e</span> <span class="st">&quot;&gt;&quot;</span> 12s_04_alberta_curated_clean.fasta <span class="op">&gt;</span> 12s_curated_header.txt</span>
<span id="cb25-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb25-2" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&#39;s/;/\t/g&#39;</span> 12s_curated_header.txt <span class="op">&gt;</span> header4rdp.txt</span>
<span id="cb25-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb25-3" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">&#39;s/&gt;//g&#39;</span> header4rdp.txt </span></code></pre></div>
<p>Add header</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb26-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb26-1" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-i</span> <span class="st">$&#39;1 i</span><span class="dt">\\\n</span><span class="st">ID</span><span class="dt">\t</span><span class="st">cellularOrganisms</span><span class="dt">\t</span><span class="st">Superkingdom</span><span class="dt">\t</span><span class="st">Kingdom</span><span class="dt">\t</span><span class="st">Phylum</span><span class="dt">\t</span><span class="st">Class</span><span class="dt">\t</span><span class="st">Order</span><span class="dt">\t</span><span class="st">Family</span><span class="dt">\t</span><span class="st">Genus</span><span class="dt">\t</span><span class="st">Species&#39;</span> header4rdp.txt</span></code></pre></div>
<p><strong>Prepare sequence file</strong></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb27-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb27-1" tabindex="-1"></a><span class="fu">sed</span> <span class="st">&#39;s/;/\t/g&#39;</span> 12s_04_alberta_curated_clean.fasta <span class="op">&gt;</span> seq4rdp.fasta</span></code></pre></div>
<p><strong>Train</strong></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb28-1"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb28-1" tabindex="-1"></a><span class="ex">python2</span> lineage2taxTrain.py header4rdp.txt <span class="op">&gt;</span> ready4train_tax.txt</span>
<span id="cb28-2"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb28-2" tabindex="-1"></a><span class="ex">python2</span> addFullLineage.py header4rdp.txt seq4rdp.fasta <span class="op">&gt;</span> ready4train_seqs.fasta </span>
<span id="cb28-3"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb28-3" tabindex="-1"></a></span>
<span id="cb28-4"><a href="generating-custom-database-of-the-12s-marker-gene.html#cb28-4" tabindex="-1"></a><span class="ex">java</span> <span class="at">-Xmx10g</span> <span class="at">-jar</span> ~/rdp_classifier_2.14/dist/classifier.jar train <span class="at">-o</span> 12s_training_files <span class="at">-s</span> ready4train_seqs.fasta <span class="at">-t</span> ready4train_tax.txt</span></code></pre></div>
<p>Output is a directory specified by the parameter -o which should contain the following files :</p>
<ul>
<li>bergeyTrainingTree.xml</li>
<li>genus_wordConditionalProbList.txt</li>
<li>logWordPrior.txt</li>
<li>wordConditionalProbIndexArr.txt</li>
</ul>
<p>Move into this newly created directory and create the file <code>rRNAClassifier.properties</code> with the following text :</p>
<pre><code># Sample ResourceBundle properties file
bergeyTree=bergeyTrainingTree.xml

probabilityList=genus_wordConditionalProbList.txt

probabilityIndex=wordConditionalProbIndexArr.txt

wordPrior=logWordPrior.txt

classifierVersion=RDP Naive Bayesian rRNA Classifier Version 2.14</code></pre>
</div>
<div id="notes-on-resolution-blindspot" class="section level2 unnumbered hasAnchor">
<h2>Notes on resolution blindspot<a href="generating-custom-database-of-the-12s-marker-gene.html#notes-on-resolution-blindspot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>Cottus</strong></p>
<p>Sequence :</p>
<p><code>CACCGCGGTTATACGAGAGGCCCAAGTTGACAAACACCGGCGTAAAGCGTGGTTAAGTTAAAAATCGTACTAAAGCCAAACATCTTCAAGACTGTTATACGTAACCGAAGACAGGAAGTTCAACCACGAAAGTCGCTTTATCTGATCTGAATCCACGAAAGCTAAGGAA</code></p>
<p>Which was exactly similar between the following taxa :</p>
<ul>
<li>NC 068673 […] Cottus cognatus</li>
<li>NC 028277 […] Cottus bairdii</li>
</ul>
<p>These were all removed from the database and replaced with a single copy of the trimmed sequence with the following sequence header :</p>
<blockquote>
<p>NC_028277:NC_068673;cellularOrganisms;Eukaryota;Metazoa;Chordata;Actinopteri;Perciformes;Cottidae;Cottus;Cottus_bairdii:cognatus`
CCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCAAGTTGACAAACACCG
GCGTAAAGCGTGGTTAAGTTAAAAATCGTACTAAAGCCAAACATCTTCAAGACTGTTAT
ACGTAACCGAAGACAGGAAGTTCAACCACGAAAGTCGCTTTATCTGATCTGAATCCACG
AAAGCTAAGGAACAAACTGGGATTAGATACCCCACTATG</p>
</blockquote>
<p><strong>Salvelinus</strong></p>
<p>Sequence :</p>
<p><code>CACCGCGGTTATACGAGAGGCCCTAGTTGATAACTACCGGCGTAAAGAGTGGTTACGGAAAAATGTTTAATAAAGCCGAACACCCCCTCAGCCGTCATACGCACCTGGGGGCACGAAGACCTACTGCGAAAGCAGCTTTAATTGTACCCGAACCCACGACAGCTACGACA</code></p>
<p>Which was exactly similar between the following taxa :</p>
<ul>
<li>NC 000861 […] Salvelinus alpinus</li>
<li>NC 037502 […] Salvelinus malma</li>
</ul>
<p>These were all removed from the database and replaced with a single copy of the trimmed sequence with the following sequence header :</p>
<blockquote>
<p>NC_000861:NC_037502;cellularOrganisms;Eukaryota;Metazoa;Chordata;Actinopteri;Salmoniformes;Salmonidae;Salvelinus;Salvelinus_alpinus:malma
CCGGTAAAACTCGTGCCAGCCACCGCGGTTATACGAGAGGCCCTAGTTGATAACTACCG
GCGTAAAGAGTGGTTACGGAAAAATGTTTAATAAAGCCGAACACCCCCTCAGCCGTCAT
ACGCACCTGGGGGCACGAAGACCTACTGCGAAAGCAGCTTTAATTGTACCCGAACCCAC
GACAGCTACGACACAAACTGGGATTAGATACCCCACTATG</p>
</blockquote>
<p><strong>Salmo</strong></p>
<p>There were 2 copies of the same 12S sequences for Salmo_trutta (NC_024032 and NC_010007), only sequence NC_024032 was kept</p>
</div>
<div id="assessing-hybrids" class="section level2 unnumbered hasAnchor">
<h2>Assessing hybrids<a href="generating-custom-database-of-the-12s-marker-gene.html#assessing-hybrids" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If one of the specie has not been found in Alberta the taxa and it’s sequence were remove. List of remove hybrids :</p>
<ul>
<li>Carassius_auratus_x_Cyprinus_carpio</li>
<li>Carassius_auratus_x_Cyprinus_carpio_x_Carassius_cuvieri</li>
<li>Cyprinus_carpio_wuyuanensis_x_Carassius_auratus</li>
</ul>
<p>Upon researching if Megalobrama amblycephala have been found in Canada and Alberta I found no evidence of such cases. Therefore all the hybrid sequences were removed.</p>
<ul>
<li>NC 013431 […] Carassius auratus x Megalobrama amblyceph ala triploid hybrid</li>
<li>NC 035976 […] Carassius auratus x Megalobrama amblycephala x Carassius cuvieri</li>
<li>NC 013430 […] Carassius auratus x Megalobrama amblycephala tetraploid hybrid</li>
<li>NC 012980 […] natural gynogenetic Carassius auratus red var.</li>
</ul>
<p>Upon researching if the hybrid has been found in Alberta / Canada I found no evidence of the such. The article associated with this genome (<a href="https://doi.org/10.1080/23802359.2019.1574636" class="uri">https://doi.org/10.1080/23802359.2019.1574636</a>) states that this hybrid was developed in Tonghua, Jilin Province, Republic of China. I therefore removed the hybrid sequence.</p>
<ul>
<li>NC 042195 […] Salvelinus fontinalis x Salvelinus malma</li>
</ul>

</div>
</div>
<hr>
<center>
  <div class="footer">
      Karine Villeneuve
      <p : id="date"></p>
<script>
  // Get current date and time
//  var now = new Date();
//  var date = now.toLocaleString();
//  n =  new Date();
//  y = n.getFullYear();
//  m = n.getMonth();
//  d = n.getDate();
//  document.getElementById("date").innerHTML = d + "/" + m + "/" + y;
  n = new Date().toDateString()
  document.getElementById("date").innerHTML = n
  // Insert date and time into HTML
//  document.getElementById("date").innerHTML = date;
</script>
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="evaluate-performance-of-databases.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="generating-custom-database.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/karinevilleneuve/fish_eDNA/master/05_custom_database_12S.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
